### PR TITLE
[11.0]  account credit control : Migration to 11.0

### PR DESCRIPTION
**Pull request by kenvac**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please add #571 to this PR :wink:

**Comment by pedrobaeza** [comment]
I'm seeing that you haven't follow migration procedure for keeping commit history: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0

Please do it that way, and then you'll get automatically also the code for #571

**Comment by kenvac** [comment]
@pedrobaeza Apologies for not following the Migration procedure (I don't know how I skip that part of documentation).
I have now a separate branch with all the changes from #571 along with this one. 

Is there a way to change the branch on PR to kenvac:11.0-mig-account_credit_control?
if not then we need to close this PR and create a new one. 

Let me know your thoughts. 

Regards,
Kinner

**Comment by pedrobaeza** [comment]
No, you have to push with `-f` option your other branch to the one with this name. You can do in local:
```
git branch -D 11.0-account_credit_control # make sure this PR branch name doesn't exist
git checkout 11.0-mig-account_credit_control # go to the branch with the correct code
git checkout -b 11.0-account_credit_control # create a branch with the name of this PR branch out of the correct one
git push -f --set-upstream kenvac 11.0-account_credit_control # upload new contents overwriting current. kenvac is the name of the remote I suppose you have
```
and with this, this PR will be automatically updated with the correct code.

**Comment by kenvac** [comment]
Hi Guys,
Any tips on how to fix codecov/project  and ci/runbot?

**Comment by astirpe** [comment]
About Runbot, there are two warnings to be solved,
one is:

```
2018-01-22 08:56:27,652 168 INFO openerp_test odoo.modules.loading: loading account_credit_control/report/report.xml
2018-01-22 08:56:27,655 168 WARNING openerp_test odoo.models: ir.actions.report.create() includes unknown fields: auto
```
and for that you should check in your code, it says that you are trying to create a record with a field `auto` which is however missing in its related model.

The second warning is:

```
2018-01-22 08:56:25,706 168 WARNING openerp_test odoo.addons.sms.wizard.send_sms: The `phonenumbers` Python module is not installed. Try: pip install phonenumbers.
```

but I guess this is a generic Runbot issue that should be addressed to the Maintainers.


Regarding the codecov, I'm not sure why it fails.

File: account_credit_control/report/report.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <report
+            id="credit_control_summary"
+            model="credit.control.communication"
+            string="Credit Control Summary"
+            report_type="qweb-pdf"
+            name="account_credit_control.report_credit_control_summary"
+            file="account_credit_control.report_credit_control_summary"
+            auto="False"
```
Comment (astirpe) [inline account_credit_control/report/report.xml]: Remove this line: `auto="False"`

File: account_credit_control/README.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: https://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+==============
+Credit Control
+==============
+
```
Comment (astirpe) [inline account_credit_control/README.rst]: It would be nice to have a minimum description of the module.

File: account_credit_control/README.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+Installation
+============
+
+Just install it
```
Comment (astirpe) [inline account_credit_control/README.rst]: Remove the "Installation" paragraph, it doesn't say anything useful...

File: account_credit_control/README.rst [hunk 4]
```diff
@@ -0,0 +1,84 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (astirpe) [inline account_credit_control/README.rst]: Use the latest [template](https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst).

File: account_credit_control/models/credit_control_line.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                            readonly=True,
+                            states={'draft': [('readonly', False)]})
+
+    state = fields.Selection([('draft', 'Draft'),
+                              ('ignored', 'Ignored'),
+                              ('to_be_sent', 'Ready To Send'),
+                              ('sent', 'Done'),
+                              ('error', 'Error'),
+                              ('email_error', 'Emailing Error')],
+                             'State',
```
Comment (astirpe) [inline account_credit_control/models/credit_control_line.py]: You can remove this line: `'State',`

File: account_credit_control/models/credit_control_line.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                             default='draft',
+                             help="Draft lines need to be triaged.\n"
+                                  "Ignored lines are lines for which we do "
+                                  "not want to send something.\n"
+                                  "Draft and ignored lines will be "
+                                  "generated again on the next run.")
+
+    channel = fields.Selection([('letter', 'Letter'),
+                                ('email', 'Email')],
+                               string='Channel',
```
Comment (astirpe) [inline account_credit_control/models/credit_control_line.py]: You can remove this line: `string='Channel',`

File: account_credit_control/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ 'version': '11.0.1.0.0',
+ 'author': "Camptocamp,Odoo Community Association (OCA),Okia,Access Bookings",
+ 'maintainer': 'Camptocamp',
+ 'category': 'Finance',
+ 'depends': [
+     'base',
+     'account',
+     'mail',
+ ],
+ 'website': 'https://www.camptocamp.com',
```
Comment (astirpe) [inline account_credit_control/__manifest__.py]: Website should be: `https://github.com/OCA/account-financial-tools`

File: account_credit_control/wizard/credit_control_marker_view.xml [hunk 8]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="credit_line_marker_form" model="ir.ui.view">
+        <field name="name">credit.line.marker.form</field>
+        <field name="model">credit.control.marker</field>
+        <field name="arch" type="xml">
+            <form string="Lines marker" version="7.0">
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_marker_view.xml]: All the forms of this module are written with the `version="7.0"`. They should instead be adapted to the latest form definition.

File: account_credit_control/wizard/credit_control_policy_changer.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                      ('move_id', '=', invoice.move_id.id),
+                      ('reconciled', '=', False)]
+            move_lines = selected_lines.search(domain)
+            selected_lines |= move_lines
+        return selected_lines
+
+    move_line_ids = fields.Many2many('account.move.line',
+                                     rel='credit_changer_ml_rel',
+                                     string='Move line to change',
+                                     default=_get_default_lines)
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_policy_changer.py]: Defaults can be better written with the lambda:
`default=lambda self: self._get_default_lines()`

**Comment by astirpe** [comment]
I'm trying to do a review of this module, but unfortunately Github does not allow me to review some of the files. That files are not displayed in the "Files changed" tab, probably it's because they are too many (129) for Github to be displayed.

**Comment by pedrobaeza** [comment]
I think you have to scroll down on the page for loading all files, but it's more comfortable to review only migration commit.

File: account_credit_control/wizard/credit_control_communication.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        return emails
+
+    @api.multi
+    def _generate_report(self):
+        """ Will generate a report by inserting mako template
+        of related policy template
+
+        """
+        report_name = 'account_credit_control.report_credit_control_summary'
+        return self.env['report'].get_pdf(self, report_name)
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_communication.py]: This part is still to be ported. Method `get_pdf()` and model `report` are not anymore in Odoo 11.0

File: account_credit_control/wizard/credit_control_printer.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+        lines = self._get_lines(self.line_ids, self._credit_line_predicate)
+
+        comms = comm_obj._generate_comm_from_credit_lines(lines)
+
+        if self.mark_as_sent:
+            comms._mark_credit_line_as_sent()
+
+        report_name = 'account_credit_control.report_credit_control_summary'
+        report_obj = self.env['report'].with_context(active_ids=comms.ids)
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_printer.py]: This part is still to be ported. Model `report` is not anymore in Odoo 11.0

File: account_credit_control/models/credit_control_run.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        id1='run_id', id2='policy_id',
+        string='Policies',
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+        default=_get_policies,
+    )
+    report = fields.Html(string='Report', readonly=True, copy=False)
+    state = fields.Selection([('draft', 'Draft'),
+                              ('done', 'Done')],
+                             string='State',
```
Comment (astirpe) [inline account_credit_control/models/credit_control_run.py]: You can remove `string='State',`

File: account_credit_control/models/credit_control_line.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+                           related='policy_level_id.level',
+                           store=True,
+                           readonly=True)
+
+    manually_overridden = fields.Boolean(string='Manually overridden')
+
+    run_id = fields.Many2one(comodel_name='credit.control.run',
+                             string='Source')
+
+    manual_followup = fields.Boolean(string='Manual Followup')
```
Comment (astirpe) [inline account_credit_control/models/credit_control_line.py]: You can remove `string='Manual Followup'`

**Review by astirpe — CHANGES_REQUESTED** [review]
This is a partial code review. Some parts of code regarding the reporting are still to be ported to V11. See my comments.

**Comment by grindtildeath** [comment]
@kenvac Are you planning to continue working on this migration ?

**Comment by kenvac** [comment]
Hi @grindtildeath,

I am going to make those recommend changes tonight and will push it.

@astirpe thanks for the review

File: account_credit_control/wizard/credit_control_communication.py [hunk 14]
```diff
@@ -208,7 +203,8 @@ def _generate_report(self):
 
         """
         report_name = 'account_credit_control.report_credit_control_summary'
-        return self.env['report'].get_pdf(self, report_name)
+        report = self.env.ref(report_name)
+        return report.render_qweb_pdf()
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_communication.py]: Thank you @kenvac ! 
I'm wondering if the method `def _generate_report()` is really necessary. It is not called from the rest of the module and it is not present in the standard Odoo also.

File: .vscode/database.json [hunk 15]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (grindtildeath) [inline .vscode/database.json]: Please remove this file.

File: account_credit_control/README.rst [hunk 16]
```diff
@@ -0,0 +1,83 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+==============
+Credit Control
```
Comment (grindtildeath) [inline account_credit_control/README.rst]: s/Credit Control/Account Credit Control

File: account_credit_control/wizard/credit_control_printer.py (hunk 11 above)
Comment (grindtildeath) [reply account_credit_control/wizard/credit_control_printer.py]: You can do :
```
report_name = 'account_credit_control.report_credit_control_summary'
report_obj = self.env['ir.actions.report']._get_report_from_name(report_name)
return report_obj.report_action(comms)
```

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Almost good, thx !
I think you can remove scenarios folder, and I'm wondering if we should not get rid of these api.returns decorators.

**Comment by grindtildeath** [comment]
Would be also nice to bring in this : https://github.com/OCA/account-financial-tools/pull/588

File: account_credit_control/views/credit_control_line.xml [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+            </search>
+        </field>
+    </record>
+
+    <record id="credit_control_line_tree" model="ir.ui.view">
+        <field name="name">credit.control.line.tree</field>
+        <field name="model">credit.control.line</field>
+        <field name="arch" type="xml">
+            <tree editable="bottom" string="Control Credit Lines"
+                  colors="green:state == 'sent';red:state in ('error', 'email_error');">
```
Comment (astirpe) [inline account_credit_control/views/credit_control_line.xml]: The `colors="..."` syntax is no longer working in Odoo 11. Please use the `decoration-` syntax.

Comment (pedrobaeza) [reply account_credit_control/views/credit_control_line.xml]: Was it working in v10? We can add this to the migration guide (https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0)

Comment (astirpe) [reply account_credit_control/views/credit_control_line.xml]: I didn't check whether is working on V10, but the `decoration-` mechanism replaces the `colors=` mechanism in the standard Odoo V10.
Maybe is better to add it in both the migration guides: Migration-to-version-10.0 and Migration-to-version-11.0.

Comment (pedrobaeza) [reply account_credit_control/views/credit_control_line.xml]: Can you please elaborate the text with the details as I don't know them? And thanks for the info! TIL

Comment (astirpe) [reply account_credit_control/views/credit_control_line.xml]: @pedrobaeza we may get some inspiration from the description of this PR: https://github.com/OCA/web/pull/839

**Comment by kenvac** [comment]
@grindtildeath the scenarios folder is there unless you mean something different and I am not sure about api.returns decoration comment. 

Could anyone please tell me how to fix codecov ?

**Comment by kenvac** [comment]
Bump!!
I've added a test to increase coverage. It should pass codecov threshold now thought can't see it.

Is there any other work to be done here please?

**Comment by grindtildeath** [comment]
@kenvac Can you also bring the fix from #588 ?
Otherwise looks good to me ! 👍 

And it would be very much appreciated if you could test and review #607 😉

**Comment by kenvac** [comment]
@grindtildeath much appreciated for the review.
Can we first merge this PR first and then create new PR for changes related to PR #588 as it is still open?

I am open to work on new PR related to #588

**Comment by kenvac** [comment]
@pedrobaeza 
Can this be merge please?

**Comment by astirpe** [comment]
@kenvac as you can notice, the review process for each PR can take a long time. If you directly include #588 in this actual PR, we will save considerable time. We are almost there :)

**Comment by kenvac** [comment]
I added few tests and increase codecov to 81.42% but it is still failing. 
Any pointers how to fix codecov? 

BTW all pending changes are complete. Ready for merge.

**Comment by astirpe** [comment]
There are many uncovered lines in [credit_control_communication.py](https://codecov.io/gh/OCA/account-financial-tools/src/5f04e026fa872770ea71afe8857227565f1b77d8/account_credit_control/wizard/credit_control_communication.py) wizard.
You can also easily cover the `def action_cancel()` method in [account_invoice.py](https://codecov.io/gh/OCA/account-financial-tools/src/5f04e026fa872770ea71afe8857227565f1b77d8/account_credit_control/models/account_invoice.py).

**Review by astirpe — APPROVED** [review]
Almost green! 
I think -1.22% in project coverage is acceptable anyway.

**Review by grindtildeath — APPROVED** [review]

File: account_credit_control/models/credit_control_policy.py [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+        res = cr.fetchall()
+        if res:
+            return different_lines.browse([row[0] for row in res])
+        return different_lines
+
+    @api.multi
+    def check_policy_against_account(self, account):
+        """ Ensure that the policy corresponds to account relation """
+        policies = self.search([])
+        allowed = [x for x in policies
```
Comment (simahawk) [inline account_credit_control/models/credit_control_policy.py]: wouldn't be better to just search w/ a proper domain?

File: account_credit_control/models/credit_control_run.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+    def _default_policies(self):
+        return self.env['credit.control.policy'].search([])
+
+    date = fields.Date(string='Controlling Date', required=True,
+                       readonly=True,
+                       states={'draft': [('readonly', False)]})
+    policy_ids = fields.Many2many(
+        'credit.control.policy',
+        rel="credit_run_policy_rel",
+        id1='run_id', id2='policy_id',
```
Comment (simahawk) [inline account_credit_control/models/credit_control_run.py]: AFAIR these have been deprecated in favour of `column1` and `column2`

File: account_credit_control/wizard/credit_control_marker_view.xml [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+        <field name="model">credit.control.marker</field>
+        <field name="arch" type="xml">
+            <form string="Lines marker">
+                <separator string="Change the state of the selected lines"
+                           colspan="4"/>
+                <newline/>
+                <label string="Warning: you will maybe not be able to revert this operation."
+                       colspan="4"></label>
+                <newline/>
+                <group>
```
Comment (simahawk) [inline account_credit_control/wizard/credit_control_marker_view.xml]: these groups deserve names :)

**Review by simahawk — APPROVED** [review]
code review only. Just a bunch of non-blocking comments.

**Review by grindtildeath — CHANGES_REQUESTED** [review]
@kenvac Can you please cherry-pick this fix commit d941ea1b1545a88ec50a9dd876079dcb398babd4 from #618  ?

File: account_credit_control/views/res_company.xml [hunk 21]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="credit_control_company_form" model="ir.ui.view">
```
Comment (grindtildeath) [inline account_credit_control/views/res_company.xml:3]: This view should be moved to res.config.settings

File: account_credit_control/models/res_company.py [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class ResCompany(models.Model):
+    """ Add credit control parameters """
+    _inherit = 'res.company'
+
+    credit_control_tolerance = fields.Float(default=0.1)
+    # This is not a property on the partner because we cannot search
+    # on fields.property (subclass fields.function).
+    credit_policy_id = fields.Many2one('credit.control.policy',
```
Comment (grindtildeath) [inline account_credit_control/models/res_company.py:14]: These two fields should also be defined as related on res.config.settings

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Can you please move the settings to `res.config.settings` ?

You should redefine the two fields on `res.company` as related and add a new view inheriting `base.res_config_settings_view_form`.

Ping me if you need help.

**Comment by pedrobaeza** [comment]
Please include the improvement in #588 and squash adjacent "OCA Transbot..." commits and squash also by logic set/author to reduce a bit the number of commits.

**Comment by kenvac** [comment]
@pedrobaeza #588 is already in the added in previous commits :)

**Review by guewen — APPROVED** [review]
Great. Squashing the OCA transbot commits would be nice indeed

**Comment by fclementic2c** [comment]
functional tests ok 
:+1: 

1 comment, I am just wondering if we shouldn't set by default all receivables account on credit control policies...
image

**Comment by kenvac** [comment]
@fclementic2c 
Thanks for the testing. I am in favour of setting receivables by default However, I can change that not a big issue.

**Comment by pedrobaeza** [comment]
Is this finished then?

**Comment by kenvac** [comment]
@fclementic2c the receivables aren't set by default on the policies. 
I wasn't sure about this and I couldn't found any code which set receivable accounts by default. 

Could you please verify this?

**Comment by kenvac** [comment]
@pedrobaeza I believe this module is finished. 
Any comments from the members please?

**Comment by simahawk** [comment]
I'd say: let's merge it. If we need further improvements like setting defaults we can issue new PRs... we are not going to freeze it :wink: 
some commits need squash (translation mainly).

**Review by grindtildeath — APPROVED** [review]
Good for me ! :+1: 
@kenvac Thanks for your work.

**Comment by pedrobaeza** [comment]
I'll squash a bit the commit history tonight and merge. Thanks for the work.

**Pull request merged**